File: Chatbots/Python/Source/io_decks.py
from __future__ import annotations

import os
from typing import List, Tuple

from .dataModel import Card, InvalidRecord, DeckPath
from .config import ParserConfig
from .normalise import normalise_for_matching
from .tokenise import tokenise


MAX_PREVIEW_CHARS = 120


def list_deck_files(data_path: str) -> List[str]:
    """Return a sorted list of .txt files under data_path (file or directory)."""
    if not os.path.exists(data_path):
        raise ValueError(f"Data path not found: {data_path}")

    if os.path.isfile(data_path):
        return [data_path] if data_path.lower().endswith(".txt") else []

    files: List[str] = []
    for root, _directories, file_names in os.walk(data_path):
        for file_name in file_names:
            if file_name.lower().endswith(".txt"):
                files.append(os.path.join(root, file_name))
    files.sort()
    return files


def split_deck_path(raw: str, separator: str) -> DeckPath:
    """Split a deck path string into a DeckPath tuple."""
    parts = [segment.strip() for segment in raw.split(separator)]
    return tuple(segment for segment in parts if segment)


def preview_raw_line(raw_line: str) -> str:
    """Return a truncated preview of a raw line for error logs."""
    raw_line = (raw_line or "").rstrip("\n\r")
    return raw_line[:MAX_PREVIEW_CHARS]


def parse_record_fields(raw_line: str) -> List[str]:
    """Split a raw line into tab-separated fields."""
    return raw_line.split("\t")


def validate_record(parts: List[str]) -> Tuple[bool, str]:
    """Check if a record has the expected structure and non-empty fields."""
    if len(parts) < 5:
        return False, "too few columns (<5)"
    if len(parts) > 6:
        return False, "too many columns (>6) or TAB inside a field"
    guid = parts[0].strip()
    question_html = parts[3].strip()
    answer_html = parts[4].strip()
    if not guid:
        return False, "empty guid"
    if not question_html:
        return False, "empty question"
    if not answer_html:
        return False, "empty answer"
    return True, ""


def read_deck_file(
    file_path: str,
    parser_config: ParserConfig,
    stopwords: set[str],
) -> Tuple[List[Card], List[InvalidRecord]]:
    """Read one deck file and return valid Cards and InvalidRecords."""
    cards: List[Card] = []
    invalid_records: List[InvalidRecord] = []

    try:
        with open(file_path, "r", encoding="utf-8") as handle:
            line_number = 0
            for raw_line in handle:
                line_number += 1
                if not raw_line or raw_line.startswith("#"):
                    continue

                parts = parse_record_fields(raw_line)
                is_valid, reason = validate_record(parts)
                if not is_valid:
                    invalid_records.append(
                        InvalidRecord(
                            file_path=file_path,
                            line_number=line_number,
                            reason=reason,
                            raw_line_preview=preview_raw_line(raw_line),
                        )
                    )
                    continue

                guid = parts[0].strip()
                deck_path_raw = parts[2].strip()
                question_raw = parts[3].rstrip("\n\r")
                answer_raw = parts[4].rstrip("\n\r")
                tags_raw = parts[5].strip() if len(parts) == 6 else ""

                deck_path = split_deck_path(deck_path_raw, parser_config.topic_separator)

                question_text = normalise_for_matching(question_raw, parser_config)
                answer_text = normalise_for_matching(answer_raw, parser_config)

                if not question_text or not answer_text:
                    invalid_records.append(
                        InvalidRecord(
                            file_path=file_path,
                            line_number=line_number,
                            reason="empty after normalisation",
                            raw_line_preview=preview_raw_line(raw_line),
                        )
                    )
                    continue

                tags = [tag.strip().lower() for tag in tags_raw.split(",") if tag.strip()] if tags_raw else []

                question_tokens = tokenise(question_text, stopwords, parser_config)
                question_token_count = len(question_tokens)

                cards.append(
                    Card(
                        guid=guid,
                        deck_path=deck_path,
                        question_raw=question_raw,
                        answer_raw=answer_raw,
                        question_text=question_text,
                        answer_text=answer_text,
                        tags=tags,
                        question_token_count=question_token_count,
                    )
                )
    except UnicodeDecodeError as exception:
        invalid_records.append(
            InvalidRecord(
                file_path=file_path,
                line_number=0,
                reason=f"unicode decode error: {exception}",
                raw_line_preview="",
            )
        )
    except OSError as exception:
        invalid_records.append(
            InvalidRecord(
                file_path=file_path,
                line_number=0,
                reason=f"file read error: {exception}",
                raw_line_preview="",
            )
        )

    return cards, invalid_records


def load_decks(
    data_path: str,
    parser_config: ParserConfig,
    stopwords: set[str],
) -> Tuple[List[Card], List[InvalidRecord]]:
    """Load all deck files under data_path and return combined Cards and InvalidRecords."""
    all_cards: List[Card] = []
    all_invalid_records: List[InvalidRecord] = []

    for file_path in list_deck_files(data_path):
        cards, invalid_records = read_deck_file(file_path, parser_config, stopwords)
        all_cards.extend(cards)
        all_invalid_records.extend(invalid_records)

    return all_cards, all_invalid_records

File: Chatbots/Rust/Source/src/io_decks.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};

use crate::data_model::{Card, DeckPath, InvalidRecord};
use crate::normalise::normalise_for_matching;

const TAB_DELIMITER: char = '\t';

/// Find all `.txt` files under a path. If the path is a file, return just that file.
pub fn list_deck_files<P: AsRef<Path>>(data_path: P) -> Result<Vec<PathBuf>> {
    let path_ref = data_path.as_ref();
    if path_ref.is_file() {
        return Ok(vec![path_ref.to_path_buf()]);
    }
    let mut files: Vec<PathBuf> = Vec::new();
    for entry in fs::read_dir(path_ref)
        .with_context(|| format!("Failed to read data directory {}", path_ref.display()))?
    {
        let entry = entry?;
        let entry_path = entry.path();
        if entry_path.is_file() {
            if let Some(extension) = entry_path.extension() {
                if extension.to_string_lossy().eq_ignore_ascii_case("txt") {
                    files.push(entry_path);
                }
            }
        }
    }
    files.sort();
    Ok(files)
}

/// Read a single Anki `.txt` deck file into `Card`s, collecting invalid records.
pub fn read_deck_file<P: AsRef<Path>>(file_path: P) -> Result<(Vec<Card>, Vec<InvalidRecord>)> {
    let path_buf = file_path.as_ref().to_path_buf();
    let file_content = fs::read_to_string(&path_buf)
        .with_context(|| format!("Failed to read deck file {}", path_buf.display()))?;

    let mut cards: Vec<Card> = Vec::new();
    let mut invalid_records: Vec<InvalidRecord> = Vec::new();

    for (zero_based_index, line_text) in file_content.lines().enumerate() {
        let line_number = zero_based_index + 1;

        // Skip metadata header lines
        if let Some(first_char) = line_text.chars().next() {
            if first_char == '#' {
                continue;
            }
        } else {
            // Empty line: skip
            continue;
        }

        // Expect at least 5 columns
        let columns: Vec<&str> = line_text.split(TAB_DELIMITER).collect();
        if columns.len() < 5 {
            invalid_records.push(InvalidRecord {
                file_path: path_buf.display().to_string(),
                line_number,
                reason: format!("Expected at least 5 columns, found {}", columns.len()),
                raw_line: line_text.to_string(),
            });
            continue;
        }

        let guid_text = columns[0].trim();
        let deck_path_text = columns[2].trim();
        let question_html = columns[3].trim();
        let answer_html = columns[4].trim();

        if guid_text.is_empty() {
            invalid_records.push(InvalidRecord {
                file_path: path_buf.display().to_string(),
                line_number,
                reason: "Empty GUID".to_string(),
                raw_line: line_text.to_string(),
            });
            continue;
        }
        if question_html.is_empty() || answer_html.is_empty() {
            invalid_records.push(InvalidRecord {
                file_path: path_buf.display().to_string(),
                line_number,
                reason: "Empty question or answer".to_string(),
                raw_line: line_text.to_string(),
            });
            continue;
        }

        let deck_path: DeckPath = if deck_path_text.is_empty() {
            Vec::new()
        } else {
            deck_path_text
                .split("::")
                .map(|segment| segment.trim().to_string())
                .collect()
        };

        let question_text = normalise_for_matching(question_html);
        let answer_raw = answer_html.to_string();

        let card = Card {
            guid: guid_text.to_string(),
            question_text,
            answer_raw,
            deck_path,
        };

        cards.push(card);
    }

    Ok((cards, invalid_records))
}

/// Load a directory (or single file) of decks and concatenate results.
pub fn load_decks<P: AsRef<Path>>(data_path: P) -> Result<(Vec<Card>, Vec<InvalidRecord>)> {
    let mut all_cards: Vec<Card> = Vec::new();
    let mut all_invalid_records: Vec<InvalidRecord> = Vec::new();

    let files = list_deck_files(&data_path)?;
    if files.is_empty() {
        return Ok((all_cards, all_invalid_records));
    }

    for file_path in files {
        match read_deck_file(&file_path) {
            Ok((mut cards, mut invalids)) => {
                all_cards.append(&mut cards);
                all_invalid_records.append(&mut invalids);
            }
            Err(error) => {
                all_invalid_records.push(InvalidRecord {
                    file_path: file_path.display().to_string(),
                    line_number: 0,
                    reason: format!("Unreadable file: {error}"),
                    raw_line: String::new(),
                });
            }
        }
    }

    Ok((all_cards, all_invalid_records))
}
